adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m

```diff
@@ -13,36 +13,75 @@
+    GADPangleRTBInterstitialRenderer *_interstitialRenderer;
+    /// Pangle rewarded ad wrapper.
+    GADPangleRTBRewardedRenderer *_rewardedlRenderer;
+}
 
 - (void)collectSignalsForRequestParameters:(nonnull GADRTBRequestParameters *)params
                          completionHandler:(nonnull GADRTBSignalCompletionHandler)completionHandler {
-  // TODO: Start signal generation and call completionHandler when the signal generation finishes or fails.
-  completionHandler("", nil);
+    NSString *signals = [BUAdSDKManager getBiddingToken:nil];
+    if (completionHandler) {
+        completionHandler(signals,nil);
+    }
 }
 
 + (void)setUpWithConfiguration:(nonnull GADMediationServerConfiguration *)configuration
              completionHandler:(nonnull GADMediationAdapterSetUpCompletionBlock)completionHandler {
-  // TODO: Set up Pangle ad network SDK and perform any necessary prefetching or configuration work.
-  completionHandler(nil);
+    NSMutableSet *appIds = [[NSMutableSet alloc]init];
+    for (GADMediationCredentials *cred in configuration.credentials) {
+        NSString *appId = cred.settings[GADMAdapterPangleAppID];
+        if (appId.length) {
+            [appIds addObject:appId];
+        }
+    }
+    
+    if (appIds.count < 1) {
+        NSError *error = GADMAdapterPangleErrorWithCodeAndDescription(GADPangleErrorInvalidServerParameters, @"Pangle mediation configurations did not contain a valid App ID.");
+        completionHandler(error);
+        return;
+    }
+    
+    NSString *appID = [appIds anyObject];
+    if (appIds.count > 1) {
+        PangleLog(@"Found the following App IDs:%@. Please remove any app IDs which you are not using", appIds);
+        PangleLog(@"Configuring Pangle SDK with the app ID:%@", appID);
+    }
+    
+    BUAdSDKConfiguration *cog = [BUAdSDKConfiguration configuration];
+    cog.territory = BUAdSDKTerritory_NO_CN;
+    cog.appID = appID;
+    [BUAdSDKManager startWithAsyncCompletionHandler:^(BOOL success, NSError *error) {
+        completionHandler(error);
+    }];
 }
 
 + (GADVersionNumber)adSDKVersion {
-  // TODO: Populate `versionString` with the ad-network's SDK version in NSString format.
-  NSString *versionString = @"";
-  GADVersionNumber version = {0};
-  NSArray<NSString *> *components = [versionString componentsSeparatedByString:@"."];
-  if (components.count == 3) {
-    version.majorVersion = components[0].integerValue;
-    version.minorVersion = components[1].integerValue;
-    version.patchVersion = components[2].integerValue;
-  } else {
-    NSLog(@"Unexpected ad SDK version string: %@. Returning 0 for adSDKVersion.", versionString);
-  }
-  return version;
+    NSString *versionString = BUAdSDKManager.SDKVersion;
+    NSArray *versionComponents = [versionString componentsSeparatedByString:@"."];
+    GADVersionNumber version = {0};
+    if (versionComponents.count == 4) {
+        NSInteger four = [versionComponents[3] integerValue];
```

Comment:
This can be removed, just do 

`version.patchVersion = [versionComponents[2] integerValue]*100+[versionComponents[3] integerValue];`

below